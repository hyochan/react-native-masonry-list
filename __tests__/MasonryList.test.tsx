import React, {ReactElement} from 'react';
import {RenderAPI, fireEvent, render} from '@testing-library/react-native';
import {Text, View} from 'react-native';

import Template from '../';
import {act} from 'react-test-renderer';

let component: ReactElement;
let testingLib: RenderAPI;

const data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

describe('Rendering', () => {
  beforeAll(() => {
    component = <Template data={data} renderItem={({i}) => <View key={i} />} />;

    testingLib = render(component);
  });

  it('should render without crashing', () => {
    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should render loading view', async () => {
    component = (
      <Template
        data={data}
        renderItem={({i}) => <View key={i} />}
        LoadingView={<Text>loading</Text>}
        loading={true}
      />
    );

    testingLib = render(component);

    const loading = await testingLib.findByText(/loading/i);

    expect(loading).toBeTruthy();
  });

  describe('Empty View', () => {
    it('should render empty view', async () => {
      component = (
        <Template
          data={[]}
          renderItem={({i}) => <View key={i} />}
          ListEmptyComponent={<Text>empty</Text>}
          loading={true}
        />
      );

      testingLib = render(component);

      const empty = await testingLib.findByText(/empty/i);

      expect(empty).toBeTruthy();
    });

    it('should render functional empty view', async () => {
      component = (
        <Template
          data={[]}
          renderItem={({i}) => <View key={i} />}
          loading={true}
          ListEmptyComponent={() => <Text>functional empty</Text>}
        />
      );

      testingLib = render(component);

      const empty = await testingLib.findByText(/functional empty/i);

      expect(empty).toBeTruthy();
    });
  });

  describe('Refresh', () => {
    it('should handle `refreshing`', async () => {
      component = (
        <Template
          data={[]}
          refreshing={true}
          renderItem={({i}) => <View key={i} />}
        />
      );

      testingLib = render(component);
      expect(testingLib).toBeTruthy();
    });

    it('should trigger `onRefresh`', async () => {
      component = (
        <Template
          testID="masonry-list"
          data={[]}
          refreshing={true}
          onRefresh={() => {}}
          numColumns={3}
          renderItem={({i}) => <View key={i} />}
        />
      );

      testingLib = render(component);

      const masonryList = testingLib.getByTestId('masonry-list');

      expect(masonryList).toBeTruthy();

      act(() => {
        masonryList.props.refreshControl.props.onRefresh();
        masonryList.props.onRefresh();
      });

      expect(masonryList.props.onRefresh).toBeDefined();
    });

    it('should trigger `onRefresh` even when `onRefresh` is not provided', async () => {
      component = (
        <Template
          testID="masonry-list"
          data={[]}
          refreshing={true}
          numColumns={3}
          renderItem={({i}) => <View key={i} />}
        />
      );

      testingLib = render(component);

      const masonryList = testingLib.getByTestId('masonry-list');

      expect(masonryList).toBeTruthy();

      act(() => {
        masonryList.props.refreshControl.props.onRefresh();
      });

      expect(masonryList.props.onRefresh).toBeUndefined();
    });
  });

  describe('Scroll', () => {
    it('should scroll to bottom and call onEndReached', () => {
      const onEndReachedFn = jest.fn();

      component = (
        <Template
          testID="masonry-list"
          data={data}
          renderItem={({i}) => <View key={i} />}
          onEndReached={onEndReachedFn}
        />
      );

      testingLib = render(component);

      const masonryList = testingLib.getByTestId('masonry-list');

      expect(masonryList).toBeTruthy();

      // Scroll to bottom
      fireEvent.scroll(masonryList, {
        nativeEvent: {
          contentSize: {height: 600, width: 400},
          contentOffset: {y: 600, x: 0},
          layoutMeasurement: {height: 100, width: 100},
        },
      });

      expect(onEndReachedFn).toBeCalled();
    });

    it('should not call `onEndReached` when undefined', () => {
      component = (
        <Template
          testID="masonry-list"
          data={data}
          renderItem={({i}) => <View key={i} />}
        />
      );

      testingLib = render(component);

      const masonryList = testingLib.getByTestId('masonry-list');

      expect(masonryList).toBeTruthy();

      // Scroll to bottom
      fireEvent.scroll(masonryList, {
        nativeEvent: {
          contentSize: {height: 600, width: 400},
          contentOffset: {y: 600, x: 0},
          layoutMeasurement: {height: 100, width: 100},
        },
      });

      expect(masonryList.props.onEndReached).toBeUndefined();
    });

    it('should not call `onEndReached` when the scroll not reached close to bottom', () => {
      const onEndReachedFn = jest.fn();

      component = (
        <Template
          testID="masonry-list"
          data={data}
          renderItem={({i}) => <View key={i} />}
          onEndReached={onEndReachedFn}
        />
      );

      testingLib = render(component);

      const masonryList = testingLib.getByTestId('masonry-list');

      expect(masonryList).toBeTruthy();

      // Scroll to bottom
      fireEvent.scroll(masonryList, {
        nativeEvent: {
          contentSize: {height: 600, width: 400},
          contentOffset: {y: 200, x: 0},
          layoutMeasurement: {height: 100, width: 100},
        },
      });

      expect(onEndReachedFn).toBeCalledTimes(0);
    });
  });
});
