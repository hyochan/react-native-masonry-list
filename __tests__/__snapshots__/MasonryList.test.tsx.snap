// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render without crashing 1`] = `
<RCTScrollView
  data={
    [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ]
  }
  keyboardShouldPersistTaps="handled"
  onScroll={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={16}
  style={
    [
      {
        "alignSelf": "stretch",
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTRefreshControl />
  <View>
    <View />
    <View
      style={
        [
          {
            "flex": 1,
            "flexDirection": "row",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          {
            "flex": 0.5,
            "flexDirection": "column",
          }
        }
      >
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
      </View>
      <View
        style={
          {
            "flex": 0.5,
            "flexDirection": "column",
          }
        }
      >
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
        <View>
          <View />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
